### DISCUSSION TITLE
How to parse a qmd file in Python?

### DISCUSSION THREAD

**Discussion by kcarnold** [quarto-dev/quarto-cli#10414] [Q&A]
### Description

I'd like to parse out the headings of all of the qmd files in a repo, to make a sitewide table-of-contents for example. (Is there any existing thing that does that?)

So, grab all of the lines that start with `#` in all the `.qmd` files, right? Not so fast; most `#`-lines are code chunk comments.

Before I go writing a parser and/or chatting with pandoc, has anyone else done this already?

**Comment by mcanouil** [comment]
That's what Pandoc/Quarto does, *.i.e* parsing/reading markdown.
You can use LUA filter or any script as a post-render script.

**Reply by kcarnold** [reply]
I briefly tried using the https://pypi.org/project/pandoc/ library and it failed on some of my qmd files because it didn't understand Figure elements. panflute worked better. Here's what I came up with; I still wish I could have just done `from qmd_parser import get_headings`. I'll toss the result in as `output: asis` chunk.

```python
from pathlib import Path
import panflute as pf
import re
import yaml

slides_qmds = []
for qmd in Path('.').rglob('*.qmd'):
    with open(qmd) as f:
        if 'revealjs' in f.read():
            slides_qmds.append(qmd)

slides_qmds.sort()

def get_toc(qmd_text: str):
    '''
    extract the headings from each file
    '''
    doc = pf.convert_text(qmd_text, input_format='markdown')
    toc = []
    for elem in doc:
        if isinstance(elem, pf.Header):
            content = pf.stringify(elem.content).strip()
            if not content:
                continue
            toc.append((elem.level, pf.stringify(elem.content)))
    return toc

for source_file in slides_qmds:
    with open(source_file) as f:
        qmd_text = f.read()
    # extract the YAML header
    yaml_header = re.search(r'^---\n(.*?)\n---', qmd_text, re.DOTALL)
    if yaml_header:
        qmd_text = qmd_text.replace(yaml_header.group(0), '')
        title = yaml.safe_load(yaml_header.group(1))['title']
    else:
        title = source_file.stem

    toc = get_toc(qmd_text)
    print(f'## {title}')
    for level, text in toc:
        print(f'{"  " * (level - 1)}- {text}')
    print()
```

**Reply by mcanouil** [reply]
I was not talking about using a Python or third party script, but to use Quarto/Pandoc directly with a LUA filter to extract headings.

**Reply by cscheid** [reply]
(sorry for the late response)

Learning our Lua API and using it really will be your best bet.

You could try to use `panflute` (a Python library that helps write pandoc JSON filters in Python), but Panflute doesn't recognize a number of Quarto-specific markdown features (shortcodes, FloatRefTargets, callouts), and so it will be a little harder than usual.

I'll also note that going through pure Pandoc will not be a great fit, because Quarto uses a custom reader, and so our Markdown syntax is not precisely that of Pandoc.

**Reply by kcarnold** [reply]
Thanks. For my specific need, the Python script I posted above, using panflute, worked well enough. But for future readers of this thread, could you give an example of what that script would look like using Quarto's custom reader?

In particular, all of the examples on the website are Quarto calling a Lua script as part of a normal rendering process. What if you wanted to use the Lua API from outside of that context (e.g., another script doing a different task)? Would you just need to pretend you're rendering something, have Lua write to a file or something, and then throw away the render output?

**Reply by cscheid** [reply]
(GitHub's post-hiding features got me here, and I completely missed your `panflute` code, sorry!)

> But for future readers of this thread, could you give an example of what that script would look like using Quarto's custom reader?

Our filter infrastructure is not particularly well suited for the task of walking the entire project. I think this is a use case where your script is actually the best strategy. I'd love to figure out a way to help, but I'm going to have to think about this a little bit.

> What if you wanted to use the Lua API from outside of that context (e.g., another script doing a different task)?

You can do `quarto run file.lua`, which, under the hood, does exactly the dance you're suggesting, and you'll have access to the Quarto APIs: https://quarto.org/docs/projects/scripts.html#periodic-scripts

Unfortunately, in this case there won't be an easy way to get to all of the .qmd files, open each of them, and so on.

Although I think this is a niche use case, it's one where "when you need, you _really_ need it". I'm going to open a new GitHub issue to mull over this.

Sorry I missed the original discussion (a confluence of multiple travel and covid had me offline for almost a month), this is a pretty important use case that we want to support!

**Reply by cscheid** [reply]
Tracking on #10539.
